src/librbd/cache/ReplicatedWriteLogInternal.h

```diff
@@ -0,0 +1,627 @@
+  }
+  const uint64_t get_offset_bytes() { return image_offset_bytes; }
+  const uint64_t get_write_bytes() { return write_bytes; }
+  friend std::ostream &operator<<(std::ostream &os,
+				  const WriteLogPmemEntry &entry) {
+    os << "entry_valid=" << (bool)entry.entry_valid << ", "
+       << "sync_point=" << (bool)entry.sync_point << ", "
+       << "sequenced=" << (bool)entry.sequenced << ", "
+       << "has_data=" << (bool)entry.has_data << ", "
+       << "discard=" << (bool)entry.discard << ", "
+       << "writesame=" << (bool)entry.writesame << ", "
+       << "sync_gen_number=" << entry.sync_gen_number << ", "
+       << "write_sequence_number=" << entry.write_sequence_number << ", "
+       << "image_offset_bytes=" << entry.image_offset_bytes << ", "
+       << "write_bytes=" << entry.write_bytes << ", "
+       << "ws_datalen=" << entry.ws_datalen << ", "
+       << "entry_index=" << entry.entry_index;
+    return os;
+  };
+};
+
+static_assert(sizeof(WriteLogPmemEntry) == 64);
+
+struct WriteLogPoolRoot {
+  union {
+    struct {
+      uint8_t layout_version;    /* Version of this structure (RWL_POOL_VERSION) */
+    };
+    uint64_t _u64;
+  } header;
+  TOID(struct WriteLogPmemEntry) log_entries;   /* contiguous array of log entries */
+  uint64_t pool_size;
+  uint64_t flushed_sync_gen;     /* All writing entries with this or a lower
+				  * sync gen number are flushed. */
+  uint32_t block_size;		 /* block size */
+  uint32_t num_log_entries;
+  uint32_t first_free_entry;     /* Entry following the newest valid entry */
+  uint32_t first_valid_entry;    /* Index of the oldest valid entry in the log */
+};
+
+typedef ReplicatedWriteLog<ImageCtx>::Extents Extents;
+
+class GenericLogEntry {
+public:
+  WriteLogPmemEntry ram_entry;
+  WriteLogPmemEntry *pmem_entry = nullptr;
+  uint32_t log_entry_index = 0;
+  bool completed = false;
+  GenericLogEntry(const uint64_t image_offset_bytes = 0, const uint64_t write_bytes = 0)
+    : ram_entry(image_offset_bytes, write_bytes) {
+  };
+  virtual ~GenericLogEntry() { };
+  GenericLogEntry(const GenericLogEntry&) = delete;
+  GenericLogEntry &operator=(const GenericLogEntry&) = delete;
+  virtual unsigned int write_bytes() = 0;
+  bool is_sync_point() { return ram_entry.is_sync_point(); }
+  bool is_discard() { return ram_entry.is_discard(); }
+  bool is_writesame() { return ram_entry.is_writesame(); }
+  bool is_write() { return ram_entry.is_write(); }
+  bool is_writer() { return ram_entry.is_writer(); }
+  virtual std::ostream &format(std::ostream &os) const {
+    os << "ram_entry=[" << ram_entry << "], "
+       << "pmem_entry=" << (void*)pmem_entry << ", "
+       << "log_entry_index=" << log_entry_index << ", "
+       << "completed=" << completed;
+    return os;
+  };
+  friend std::ostream &operator<<(std::ostream &os,
+				  const GenericLogEntry &entry) {
+    return entry.format(os);
+  }
+};
+
+} // namespace rwl
+
+class ReplicatedWriteLogInternal {
```

Comment:
IIRC, it's to keep some PMDK macros that use m_pool_root out of ReplicatedWriteLog.h. That helps contain what has to include the PMDK headers.

The real reason ReplicatedWriteLogInternal.h exists is that the ReplicatedWriteLog.[ch] files were huge when I was working on this alone. That was always a bad idea, but never got in my way. I had to split this up when the Ceph team took this over because it was too big for our Internal gitlab code review tool.

Reply:
We removed ReplicatedWriteLogInternal